test-data/unit/check-dataclasses.test

```diff
@@ -2209,17 +2217,33 @@ a: Any
 a2 = replace(a)
 reveal_type(a2)  # N: Revealed type is "Any"
 
+[builtins fixtures/tuple.pyi]
+
 [case testReplaceNotDataclass]
 from dataclasses import replace
 
-replace(5)  # E: Argument 1 to "replace" has incompatible type "int"; expected a dataclass
+replace(5)  # E: Value of type variable "_DataclassT" of "replace" cannot be "int"
 
 class C:
     pass
 
-replace(C())  # E: Argument 1 to "replace" has incompatible type "C"; expected a dataclass
+replace(C())  # E: Value of type variable "_DataclassT" of "replace" cannot be "C"
 
-replace(None)  # E: Argument 1 to "replace" has incompatible type "None"; expected a dataclass
+replace(None)  # E: Value of type variable "_DataclassT" of "replace" cannot be "None"
+
+[builtins fixtures/tuple.pyi]
+
+[case testReplaceIsDataclass]
+from dataclasses import is_dataclass, replace
+
+def f(x: object) -> None:
+  # error before type-guard
+  y = replace(x)  # E: Value of type variable "_DataclassT" of "replace" cannot be "object"
+  # no error after type-guard
+  if is_dataclass(x) and not isinstance(x, type):
```

Comment:
Do we need to test just `if is_dataclass(x):`?

Reply:
It's because type is an object, and `is_dataclass` on an object either means it's a dataclass instance _or_ type.

I'll improve the test to clarify that.

Reply:
1a538d4306ee6b5d1903720a99c18edcf4326b03